### DISCUSSION TITLE
Is it possible to supply the EntityLinker component with a Child Class of KnowledgeBase?

### DISCUSSION THREAD

**Discussion by coltonflowers1** [explosion/spaCy#12289] [Help: Coding & Implementations] [answered]
I am trying to subclass the `KnowledgeBase` class, and then provide that to the `EntityLinker` component. However, it looks like the `EntityLinker` component's  `from_disk` method contains the following:

```
deserialize["kb"] = lambda p: self.kb.from_disk(p)
```
I'm used to having `from_disk` be a class method, but here it is called by the object, so it's assuming that `self.kb` has already been initialized by the time this function gets called. However, the only time I see `self.kb` getting initialized at inference time(i.e. not in the `initialize` method)  is in the `__init__` method of `EntityLinker` in this line:
```
 self.kb = empty_kb(entity_vector_length)(self.vocab)
```
`empty_kb` actually returns an `InMemoryLookupKB` object, which leads me to believe that `EntityLinker` is using this subclass's `from_disk` method and would not use the `KnowledgeBase` subclass that was supplied at training time using a custom `kb_loader ` method.

Am I missing something here? Is it the EntityLinker component not meant to be supplied with a `kb_loader` callable that returns something other than an `InMemoryLookupKB` object?

**Comment by svlandeg** [comment]
Hi @coltonflowers1. Thanks for the detailed explanation - your analysis is pretty much correct and is leaving me wondering why I implemented this as such. It's originating from [this PR](https://github.com/explosion/spaCy/pull/6216/) which actually has an explicit test for the KB being the "standard" type after IO. At the time this was `KnowledgeBase`, since 3.5 it would be `InMemoryLookupKB` as you mention, and as you can see in the current version of this test here: https://github.com/explosion/spaCy/blob/master/spacy/tests/serialize/test_serialize_kb.py#L142

This doesn't seem right - I'll look into it more and get back to you.

**Comment by svlandeg** [comment, marked as answer]
We're working on a fix for this here: https://github.com/explosion/spaCy/pull/12320

**Reply by svlandeg** [reply]
Basically from 3.5.1 onwards, you'll be able to modify the behaviour of what was previously `empty_kb()` by defining this function in the config:
```
[components.entity_linker.generate_empty_kb]
    @misc = "kb_test.CustomEmptyKB.v1"
```
As example usage, you can have a look at the `test_serialize_subclassed_kb` test [here](https://github.com/explosion/spaCy/blob/master/spacy/tests/serialize/test_serialize_kb.py).
